### ISSUE TITLE
Problem in reading csv data (inconsistent dimensions)

### ISSUE THREAD

**Issue by danesh-d** [mlpack/mlpack#3593]
Hi all,
I am using two very simple csv files where the first one contains training data (x) and the other one which has only one column, the target data that is intended to be predicted (y). I use the code below to read the data but when I run the code, I get an "out of bound" runtime error. I changed all parameters in RandomForest class initialization to 1 and then I got a segmentation fault error.

The other problem I see is, when I dump the matrix of training data and the vector of the target data, the elements are somehow transposed and not shown in the original order. I am wondering if I have organized and read my training data correctly?

```
int main() {
   mat data;
   data::Load("peec_train.csv", data);
   Row<size_t> labels;
   data::Load("peec_target.csv", labels);

   labels -= 1;

   mat trainDataset, testDataset;
   Row<size_t> trainLabels, testLabels;
   data::Split(data, labels, trainDataset, testDataset, trainLabels, testLabels, 0.1);

   // Create the RandomForest object and train it on the training data.
   RandomForest<> r(trainDataset,
                    trainLabels,
                    7,
                    10,
                    3);
}
```

Here is my training data:

0,120,4032,4008,4008,4032,0,0,6,63383,838,0.0,10
0,120,4032,4008,4008,4032,1,0,6,63383,838,0.0,22
0,120,4032,4008,4008,6016,1,1,6,63383,838,0.0,19
0,120,0,4008,4008,4008,1,0,6,63383,836,0.0,12
0,120,6060,10524,10524,15042,0,1,6,63383,6633,0.0,103
0,120,6060,10524,10524,15042,1,1,6,63383,6633,0.0,138
1,350,12060,21024,21024,21024,0,0,6,63383,5250,0.0,7
1,350,12060,21024,21024,21024,0,1,6,63383,5250,0.0,10

Here is my target data:

10
22
19
12
103
138
7
10

I know my data set is very limited but I want to test to make sure that I am using the library correctly, then extend the data set.

My last question is, whether mlpack supports a reinforcement learning scheme or something like online learning? I mean, something like, using a trained ML model to generate prediction, then by having the ground truth, if the error is higher than a threshold, then train the ML model with the ground truth data.

Thanks for your help,

Dan

**Comment by shrit** [comment]
Hi @danesh-d can you share the csv file here?

**Comment by rcurtin** [comment]
Hey there @danesh-d, the code looks pretty much correct but there is one major issue: the labels need to be between `0` and `numClasses - 1`.  In your case, the target data takes large values that are greater than `numClasses`.  We are currently working on documentation to clarify these issues---for the random forest, you can see this is pointed out [here](https://www.ratml.org/misc/mlpack-markdown-doc/random_forest.html#constructors) (see the table of constructor parameters).

mlpack actually does have support to do this mapping, with the `data::NormalizeLabels()` function.  Here is an adaptation of your code to map the labels to `[0, numClasses)`, and then map back some predictions:

```c++
#include <mlpack.hpp>

using namespace arma;
using namespace mlpack;

int main()
{
  mat data;
  data::Load("peec_train.csv", data);
  Row<size_t> labels;
  data::Load("peec_target.csv", labels);

  // The labels must be between 0 and the number of classes.
  // We can use data::NormalizeLabels() for this.
  Col<size_t> mappings;
  Row<size_t> mappedLabels;
  data::NormalizeLabels(labels, mappedLabels, mappings);

  mat trainDataset, testDataset;
  Row<size_t> trainLabels, testLabels;
  data::Split(data, mappedLabels, trainDataset, testDataset, trainLabels, testLabels, 0.1);

  RandomForest<> r(trainDataset, trainLabels, 7, 10, 3);

  // Make some predictions, then map them back to the original classes.
  Row<size_t> predictions;
  r.Classify(trainDataset, predictions);
  predictions.print("Predictions before class unmapping:");

  Row<size_t> unmappedPredictions;
  data::RevertLabels(predictions, mappings, unmappedPredictions);
  unmappedPredictions.print("Predictions after class unmapping:");
}
```

When you run it, you get:

```
Predictions before class unmapping:
        4        0        1        0        5        1        1        1
Predictions after class unmapping:
          103           10           22           10          138           22           22           22
```

I know that `NormalizeLabels()` is not well documented.  That is a work in progress!  (I am actually working on exactly that today.)

Two other small comments:

 * Your labels are the last dimension of your input data, so consider using `.shed_row()` on `data` after loading it.  Otherwise the learning task is really trivial.
 * For a dataset this small, a minimum leaf size of `3` is probably too large.  (Of course, I know you are just experimenting, so you probably already know this.)

Hope that helps, let me know if I can clarify anything. :+1:

**Comment by rcurtin** [comment]
> Can it be, because my random data is not uniformly distributed and the distribution might be biased?

What have you tried to debug the issue?  It's almost certain to me based on your output that the data dimensions are not linearly independent and thus regularization is needed.  That's a standard issue with linear regression models.

> I would at least add support for famous libraries like Intel MKL

Armadillo supports MKL.  https://arma.sourceforge.net/faq.html

> C raw pointers to store data

If you knew what it was like to maintain complicated linear algebra algorithms when working directly with C pointers, you would not suggest this.  It isn't a good idea.

> Also, I would add sample code which could only create an ML mode using different algorithms and perform very simple predictions.

https://github.com/mlpack/examples

> the package is not well documented

This is something we are actively working on.
